lib/kernel/src/code.erl

```diff
@@ -248,10 +253,16 @@ load_binary(Mod, File, Bin)
 
 load_module(Mod, File, Bin, Purge) ->
     case erlang:prepare_loading(Mod, Bin) of
-        {error,_}=Error ->
-            Error;
-        Prepared ->
-            call({load_module, Prepared, Mod, File, Purge, false})
+        {error,_}=Error -> Error;
+        Prepared -> load_prepared_module(Prepared, Mod, File, Purge, false)
+    end.
+
+load_prepared_module(Prepared, Mod, File, Purge, EnsureLoaded) ->
+    case get_mode() of
```

Comment:
I don't see a guard like that in the handler for `load_module`. Are you sure this change is correct? It seems to do a lot of extra stuff regardless of mode.

Reply:
`handle(load_module)` calls `try_finish_module`, which is what would trigger this. But it seems it would only trigger this if `EnsureLoaded` is not false, so I have to at least address that. Good find.

Reply:
@michalmuskala can you please do another review? I pushed more granular changes and I think that's better now, thank you!

Reply:
I'm still not 100% convinced this is fully equivalent. The code in `code_server` did call `handle_pending_on_load` even in the embedded mode, and produced the `{error, embedded}` errors only once the on load handling was finished (and module still not loaded). Is it ok not to call that?

I think it might not be ok, in particular I can see this sequence where behaviour would differ:
1. Module `m` with an `on_laod` handler
2. Process A loads this module with `code:load_file(m)` which works in embedded mode. Code server schedules `on_load` handler to execute
3. In parallel process B calls `code:ensure_modules_loaded([m])` - this is routed to the server, which notices `on_load` callback for module `m` is pending
4. Once `on_load` callback finishes, the module is considered loaded and process B returns `ok`, while in this implementation it could return `{error, embedded}`.

Does this make sense?

Reply:
@michalmuskala that's a good example but I am not sure we want to decide this based on a race condition? If `code:ensure_modules_loaded` just happens before `load_file`, then you would get `{error, embedded}`.

I am also trying to look at how it behaves in Erlang/OTP 25, and Erlang/OTP 25 would not wait for `on_load` callbacks either: https://github.com/erlang/otp/blob/maint-25/lib/kernel/src/code.erl#L435 - but it would do a one last ditch attempt of checking if the module is loaded (but note we already perform such check earlier, so I would say the later check is a side-effect of the `ensure_loaded` call).

Reply:
Yeah, I'm not 100% sure this change in behaviour is a problem - just pointing out that it's there.
In general I find the code server code extremely hard to follow and the various conditions & invariants of different functions interacting in quite complex ways.

I wonder if, long term, it would make sense to do a major refactoring by duplicating the code for each of the various operations, so it's not generic at all, and then extracting common utilities back again - I think this might make it easier to follow what's happening at the end of such exercise.

Reply:
Ok, after further studying the code for 1h...

Another possible question to ask is: how to make `ensure_modules_loaded([Module])` and `ensure_loaded(Module)` as consistent as possible.

We will most likely have one major difference between them forever: `ensure_loaded/1` will not allow concurrent loading of the same module (and I believe we should at least document that).

There are still two other ways where they differ:

1. `ensure_loaded/1` never loads a module in embedded mode, `ensure_modules_loaded/1` skips loading only modules with `on_load` callbacks

2. `ensure_loaded/1` logs stick directories and failure to load modules, `ensure_modules_loaded/1` reports those only with `on_load` callbacks

Those differences happen because for modules without `on_load`, `ensure_modules_loaded/1` calls `finish_loading/1` which is a single `code_server` call to processes all modules at once, which does not log errors. For modules with `on_load`, it goes through the same code path as `ensure_loaded/1`. This even makes the docs inconsistent, because the docs say `ensure_modules_loaded/1` behaves the same as `load_file/1`, and that's not true, because `load_file/1` logs and `ensure_modules_loaded/1 ` mostly does not.

So the only way to make it consistent is to replace the single `finish_loading` GenServer call by N `load_module` calls. Internally, they do the same amount of work, N times. So maybe doing this change wouldn't change the performance but it is unsure.

In any case, this pull request at least makes the behaviour between `ensure_loaded/1` and `ensure_modules_loaded/1` at least a bit more consistent and back to Erlang/OTP 25.